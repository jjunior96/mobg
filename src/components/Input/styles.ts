import { lighten } from 'polished';
import styled, { css } from 'styled-components';

import { InputProps, LabelColorType } from '.';

type DisableProps = Pick<InputProps, 'isDisabled'>;

type IconContainerProps = {
  isFocused: boolean;
  isErrored?: boolean;
} & DisableProps;

type LabelProps = {
  isRequired?: boolean;
  labelColor?: LabelColorType;
} & DisableProps;

export const Container = styled.div``;

export const Label = styled.label<LabelProps>`
  ${({ theme, isRequired, isDisabled, labelColor }) => css`
    font-size: ${theme.font.sizes.xxsmall};
    margin: 0;
    padding: 0;
    line-height: 1;

    span {
      display: none;
    }

    color: ${labelColor === 'white'
      ? theme.colors.white
      : theme.colors.blue_300};

    ${isRequired &&
    css`
      span {
        display: inline;
        color: ${theme.colors.red};
      }
    `}

    ${isDisabled &&
    css`
      cursor: not-allowed;

      color: ${theme.colors.gray_200};
    `}
  `}
`;

export const IconContainer = styled.div<IconContainerProps>`
  ${({ theme, isErrored, isFocused, isDisabled }) => css`
    background: ${theme.colors.white};
    color: ${theme.colors.gray_200};

    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xsmall} 2rem;

    width: 100%;
    height: 56px;

    font-size: ${theme.font.sizes.medium};

    display: flex;
    align-items: center;

    border: 2px solid ${theme.colors.gray_200};

    cursor: default;

    &:focus-within {
      /* box-shadow: 0 0 0.5rem ${theme.colors.primary}; */
      outline: 3px solid ${lighten(0.4, theme.colors.primary)};
    }

    input {
      color: ${theme.colors.blue_300};
      background: transparent;
      outline: none;

      border: 0;

      flex: 1;

      // Remove as setas up e down nos inputs do tipo number
      &::-webkit-outer-spin-button,
      &::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
      }
      &[type='number'] {
        -moz-appearance: textfield;
      }

      /* Remove os estilos nas sugestÃµes do Chrome */
      &:-webkit-autofill,
      &:-webkit-autofill:hover,
      &:-webkit-autofill:focus {
        -webkit-text-fill-color: ${theme.colors.gray_200};
        -webkit-box-shadow: #282f3a26 inset;
        box-shadow: #282f3a26;
        transition: background-color 5000s ease-in-out 0s;
      }

      &::placeholder {
        color: ${theme.colors.gray_200};
      }

      ${isDisabled &&
      css`
        cursor: not-allowed;

        color: ${theme.colors.gray_200};
      `}
    }

    svg {
      margin-right: ${theme.spacings.xsmall};
      color: ${theme.colors.gray_200};
    }

    ${isErrored &&
    css`
      border-color: ${theme.colors.red};

      > svg {
        color: ${theme.colors.red};
      }
    `}

    ${isFocused &&
    css`
      border-color: ${theme.colors.primary};
      color: ${theme.colors.primary};

      > svg {
        color: ${theme.colors.primary};
      }
    `}


    ${isDisabled &&
    css`
      cursor: not-allowed;

      border-color: ${lighten(0.3, theme.colors.gray_200)};

      color: ${lighten(0.3, theme.colors.gray_200)};

      > svg {
        color: ${lighten(0.3, theme.colors.gray_200)};
      }
    `}
  `}
`;

export const Error = styled.div`
  ${({ theme }) => css`
    height: 2rem;
    margin: 0;
    padding: 0;
    margin-bottom: ${theme.spacings.xxsmall};

    svg {
      margin: 0;
      padding: 0;
    }

    p {
      font-size: ${theme.font.sizes.xsmall};
      color: ${theme.colors.red};
    }
  `}
`;
