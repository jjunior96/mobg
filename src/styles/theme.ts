export default {
  grid: {
    container: '130rem',
    gutter: '3.2rem'
  },
  border: {
    radius: '0.4rem'
  },
  box: {
    shadow: '0 8px 8px rgba(0, 0, 0, 0.08)'
  },
  font: {
    family: {
      primary:
        "'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
      secundary:
        "'Lexend', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif"
    },

    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xxxxsmall: '0.8rem',
      xxxsmall: '1.2rem',
      xxsmall: '1.4rem',
      xsmall: '1.6rem',
      small: '2rem',
      medium: '2.4rem',
      large: '3.2rem',
      xlarge: '4rem',
      xxlarge: '4.8rem',
      xxxlarge: '5.6rem',
      huge: '6.4rem'
    }
  },
  colors: {
    primary: '#5c63ef',
    secundary: '#d8d8e5',
    white: '#fff',
    gray_100: '#e6e6eb',
    gray_200: '#d8d8e5',
    blue_100: '#5c63ef',
    blue_200: '#5d5e7a',
    blue_300: '#292b52',
    red: '#e03f3f',
    green: '#2ea340',
    orange: '#ff8933',
    black: '#292b52'
  },
  spacings: {
    xxxsmall: '0.8rem',
    xxsmall: '1rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const;
