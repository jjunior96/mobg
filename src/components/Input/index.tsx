import { InputHTMLAttributes, useState, useCallback } from 'react';

import * as S from './styles';

export type LabelColorType = 'white' | 'black';

type Props = {
  label?: string;
  labelColor?: LabelColorType;
  isRequired?: boolean;
  error?: string;
  isDisabled?: boolean;
  icon?: React.ReactNode;
};

export type InputProps = InputHTMLAttributes<HTMLInputElement> & Props;

const Input = ({
  label,
  labelColor = 'black',
  name,
  isRequired = false,
  error,
  isDisabled = false,
  type = 'text',
  icon,
  ...rest
}: InputProps) => {
  const [isFocused, setIsFocused] = useState(false);

  const handleInputFocus = useCallback(() => {
    setIsFocused(true);
  }, []);

  const handleInputBlur = useCallback(() => {
    setIsFocused(false);
  }, []);

  return (
    <S.Container>
      <S.Label
        htmlFor={label}
        isRequired={isRequired}
        isDisabled={isDisabled}
        labelColor={labelColor}
      >
        {label} <span>*</span>
      </S.Label>
      <S.IconContainer
        isFocused={isFocused}
        isErrored={!!error}
        isDisabled={isDisabled}
      >
        {icon}
        <input
          name={name}
          type={type}
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          disabled={isDisabled}
          {...rest}
        />
      </S.IconContainer>

      {error && (
        <S.Error title={error}>
          <p>{error}</p>
        </S.Error>
      )}
    </S.Container>
  );
};

export default Input;
