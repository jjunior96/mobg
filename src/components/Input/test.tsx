import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import theme from 'styles/theme';

import Input from '.';

describe('<Input />', () => {
  it('should render the Input component', () => {
    const { container } = renderWithTheme(
      <Input name="Name" aria-label="name-input" />
    );

    expect(
      screen.getByRole('textbox', { name: /name-input/i })
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the Input component with label', () => {
    renderWithTheme(<Input label="Name" />);

    expect(screen.getByText(/name/i)).toHaveStyle({
      'font-size': '1.4rem',
      color: `${theme.colors.blue_300}`
    });
  });

  it('should render the Input component with labelColor', () => {
    renderWithTheme(<Input label="Name" labelColor="white" />);

    expect(screen.getByText(/name/i)).toHaveStyle({
      color: `${theme.colors.white}`
    });
  });

  it('should render the Input component when is required', () => {
    renderWithTheme(<Input label="Name" isRequired />);

    // Quando `isRequired` é passado, é exibido um span com "*" para
    // simbolizar que o campo é obrigatório
    expect(screen.getByText(/\*/)).toHaveStyle({
      display: 'inline',
      color: theme.colors.red
    });
  });

  it('should render the Input component when error is passed', () => {
    renderWithTheme(<Input label="Name" error="Insert a name valid" />);

    expect(screen.getByText(/Insert a name valid/i)).toBeInTheDocument();
  });

  it('should is not accessible by tab when disabled', () => {
    renderWithTheme(<Input label="Name" isDisabled />);

    const input = screen.getByRole('textbox');
    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).not.toHaveFocus();
  });
});
