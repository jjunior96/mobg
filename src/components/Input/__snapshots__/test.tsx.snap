// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render the Input component 1`] = `
.c0 {
  font-size: 1.4rem;
  margin: 0;
  padding: 0;
  line-height: 1;
  color: #292b52;
}

.c0 span {
  display: none;
}

.c1 {
  background: #fff;
  color: #d8d8e5;
  border-radius: 0.8rem;
  padding: 1.6rem 2rem;
  width: 100%;
  height: 56px;
  font-size: 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid #d8d8e5;
  cursor: default;
}

.c1:focus-within {
  outline: 3px solid #fff;
}

.c1 input {
  color: #292b52;
  background: transparent;
  outline: none;
  border: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 input::-webkit-outer-spin-button,
.c1 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c1 input[type='number'] {
  -moz-appearance: textfield;
}

.c1 input:-webkit-autofill,
.c1 input:-webkit-autofill:hover,
.c1 input:-webkit-autofill:focus {
  -webkit-text-fill-color: #d8d8e5;
  -webkit-box-shadow: #282f3a26 inset;
  box-shadow: #282f3a26;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
}

.c1 input::-webkit-input-placeholder {
  color: #d8d8e5;
}

.c1 input::-moz-placeholder {
  color: #d8d8e5;
}

.c1 input:-ms-input-placeholder {
  color: #d8d8e5;
}

.c1 input::placeholder {
  color: #d8d8e5;
}

.c1 svg {
  margin-right: 1.6rem;
  color: #d8d8e5;
}

<div
  class=""
>
  <label
    class="c0"
  >
     
    <span>
      *
    </span>
  </label>
  <div
    class="c1"
  >
    <input
      aria-label="name-input"
      name="Name"
      type="text"
    />
  </div>
</div>
`;
